crates/sui-graphql-rpc/src/types/object.rs

```diff
@@ -410,39 +460,168 @@ impl Object {
+        let vec_address = address.into_vec();
+        use objects::dsl as objects;
 
         let stored_obj: Option<StoredObject> = db
             .optional(move || {
-                let mut query = dsl::objects
-                    .filter(dsl::object_id.eq(address.clone()))
+                objects::objects
+                    .filter(objects::object_id.eq(vec_address.clone()))
                     .limit(1)
-                    .into_boxed();
+                    .into_boxed()
+            })
+            .await?;
 
-                // TODO: leverage objects_history
-                if let Some(version) = version {
-                    query = query.filter(dsl::object_version.eq(version));
-                }
+        stored_obj.map(Self::try_from).transpose()
+    }
 
-                query
+    async fn historical_object_query(
+        db: &Db,
+        address: &SuiAddress,
+        version: Option<i64>,
+        checkpoint_sequence_number: Option<i64>,
+    ) -> Result<Option<Self>, Error> {
+        let vec_address = address.into_vec();
+
+        use checkpoints::dsl as checkpoints;
+        use objects_history::dsl as history;
+        use objects_snapshot::dsl as snapshot;
+
+        let results: Option<Vec<StoredHistoryObject>> = db
+            .inner
+            .spawn_blocking(move |this| {
+                this.run_consistent_query(|conn| {
+                    // If an object was created or mutated in a checkpoint outside the current
+                    // available range, and never touched again, it will not show up in the
+                    // objects_history table. Thus, we always need to check the objects_snapshot
+                    // table as well.
+                    let mut snapshot_query = snapshot::objects_snapshot
+                        .filter(snapshot::object_id.eq(vec_address.clone()))
+                        .into_boxed();
+
+                    let mut historical_query = history::objects_history
+                        .filter(history::object_id.eq(vec_address.clone()))
+                        .order_by(history::object_version.desc())
+                        .limit(1)
+                        .into_boxed();
+
+                    // version is an exact lookup
+                    if let Some(version) = version {
+                        snapshot_query =
+                            snapshot_query.filter(snapshot::object_version.eq(version));
+                        historical_query =
+                            historical_query.filter(history::object_version.eq(version));
+                    }
+
+                    let left = snapshot::objects_snapshot
+                        .select(snapshot::checkpoint_sequence_number)
+                        .order(snapshot::checkpoint_sequence_number.desc())
+                        .limit(1);
+
+                    // if some checkpoint_sequence_number is provided, use it as the upper bound.
```

Comment:
I'm not fully following why we need this query to be honest -- maybe we can discuss when we meet today, but I would have thought that either the data is just there in the historical or snapshot data, or it is not.

When the indexer generates/updates the snapshot from the historical data, does it drop the partitions it generated the snapshot from?

Reply:
It's possible for an object to exist both in snapshot and history, this remains true after the first time an object gets snapshotted. Say snapshot is at checkpoint 100, and has single entry for an object created at that checkpoint. Later in history, the object was mutated again at checkpoint 150, the object exists in both snapshot and history.

If an object query is issued with version, then the possibilities are that the object at that version exists only in snapshot, or only in history. It should be impossible for the object at the version to be in both snapshot and history, since history's start_cp is snapshot's max_cp + 1, and history does not track any objects that were not mutated in a checkpoint. 

If an object query is issued only with checkpoint_sequence_number, then we may get results from both snapshot and history. We privilege data from history as it is more recent, and otherwise return snapshot if history is null.

Reply:
Got it, thanks for clarifying! (and as per our conversation last week, the checkpoint sequence number we're supplying here is part of the Object cursor key (which is just ID and checkpoint version).